### PR TITLE
Makes engine extensible (internally) 

### PR DESCRIPTION
**Pull request by mathetake**
This allows us to develop the new backend without introducing huge dependency into this wazero repository.
The trick here seems really cool and we can achieve the goal without modifying any internal code or exposing
any existing structure in the public surface. That is this doesn't expand any public API surface of wazero.

This supersedes #1049 and the continuation of the development on the new backend will continue in 
the new repository (https://github.com/tetratelabs/wazerolift/pull/1) (the name is temporary :D).

The below is the example of what's possible now from a user perspective ([the code is here](https://github.com/tetratelabs/wazerolift/blob/initialimpl/example/example_test.go)).

```go
package example

import (
	"context"
	_ "embed"

	"github.com/tetratelabs/wazero"
	"github.com/tetratelabs/wazerolift"
)

// addWasm was generated by the following:
//
//	wasm-tools parse testdata/add.wat -o testdata/add.wasm
//
//go:embed testdata/add.wasm
var addWasm []byte

func main() {
	config := wazero.NewRuntimeConfig()
	// Configures Cranelift backend by passing wazero.RuntimeConfig.
	wazerolift.ConfigureCranelift(config)

	ctx := context.Background()
	r := wazero.NewRuntimeWithConfig(ctx, config)
	inst, _ := r.InstantiateModuleFromBinary(ctx, addWasm)

	add := inst.ExportedFunction("add")

	result, _ := add.Call(ctx, 1, 2)
	if result[0] != 3 {
		panic("BUG")
	}
}
```

### CODE REVIEW

**Comment by mathetake** [comment]
https://github.com/golangci/golangci-lint/issues/3592 hit this

File: internal/engineext/engineext.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+// If we directly use this interface in the signature directly, Go runtime treats this and copied interface differently
+// and makes it impossible to insert the custom EngineExt, ModuleEngineExt, etc.
+type FunctionInstance interface {
+	// ModuleInstanceName returns wasm.ModuleInstance Name of this function.
+	ModuleInstanceName() string
+	// FunctionType returns the signature of this function.
+	FunctionType() (params []api.ValueType, results []api.ValueType)
+	// Index returns wasm.FunctionInstance Idx.
+	Index() Index
+}
```
Comment (mathetake) [inline internal/engineext/engineext.go:119]: these interfaces will change overtime.

File: config.go [hunk 2]
```diff
@@ -157,9 +158,16 @@ func NewRuntimeConfig() RuntimeConfig {
...
+type (
+	newEngine    func(context.Context, api.CoreFeatures, filecache.Cache) wasm.Engine
+	newEngineExt func(context.Context, api.CoreFeatures, any) engineext.EngineExt
+)
 
 type runtimeConfig struct {
+	// newEngineExt is not used by default. Only set when the external
+	// engine implementation is used. This field is manipulated via unsafe type conversion in such
+	// out-of-source engine implementation. Therefore, it is ok to keep it unexported.
+	newEngineExt          //nolint
```
Comment (mathetake) [inline config.go:170]: this is the key of the trick.

**Comment by mathetake** [comment]
We will merge after 1.0 next month.

**Comment by codefromthecrypt** [comment]
Thanks for deferring on this. I think a go.mod replace until then pointing to this branch will help!

**Comment by codefromthecrypt** [comment]
@mathetake do you mind closing this for repo hygiene purposes? I think the experiment is neat, but needn't be in the form of a PR

**Comment by mathetake** [comment]
thanks for the ping!

**Pull request closed without merging**